examples/cryptocurrency-advanced/backend/src/schema.rs

```diff
@@ -14,94 +14,106 @@
 
 //! Cryptocurrency database schema.
 
-use exonum_merkledb::{IndexAccess, ObjectHash, ProofListIndex, ProofMapIndex};
+use exonum_merkledb::{AccessExt, IndexAccessMut, ObjectHash, ProofListIndex, ProofMapIndex};
 
 use exonum::crypto::{Hash, PublicKey};
 
 use crate::{wallet::Wallet, INITIAL_BALANCE};
 
+const NOT_INITIALIZED: &str = "Cryptocurrency schema is not initialized";
+
 /// Database schema for the cryptocurrency.
 #[derive(Debug)]
-pub struct Schema<'a, T> {
-    service_name: &'a str,
+pub struct Schema<T: AccessExt> {
+    /// Map of wallet keys to information about the corresponding account.
+    pub wallets: ProofMapIndex<T::Base, PublicKey, Wallet>,
     access: T,
 }
 
-impl<'a, T> Schema<'a, T>
-where
-    T: IndexAccess,
-{
+impl<T: AccessExt> Schema<T> {
     /// Creates a new schema from the database view.
-    pub fn new(service_name: &'a str, access: T) -> Self {
-        Schema {
+    pub fn new(access: T) -> Self {
+        Self {
+            wallets: access.proof_map("wallets").expect(NOT_INITIALIZED),
             access,
-            service_name,
         }
     }
 
-    /// Returns `ProofMapIndex` with wallets.
-    pub fn wallets(&self) -> ProofMapIndex<T, PublicKey, Wallet> {
-        ProofMapIndex::new(
-            [self.service_name, ".wallets"].concat(),
-            self.access.clone(),
-        )
-    }
-
     /// Returns history of the wallet with the given public key.
-    pub fn wallet_history(&self, public_key: &PublicKey) -> ProofListIndex<T, Hash> {
-        ProofListIndex::new_in_family(
-            [self.service_name, ".wallet_history"].concat(),
-            public_key,
-            self.access.clone(),
-        )
+    pub fn wallet_history(&self, public_key: &PublicKey) -> Option<ProofListIndex<T::Base, Hash>> {
+        self.access.proof_list(("wallet_history", public_key))
     }
 
     /// Returns wallet for the given public key.
     pub fn wallet(&self, pub_key: &PublicKey) -> Option<Wallet> {
-        self.wallets().get(pub_key)
+        self.wallets.get(pub_key)
     }
 
     /// Returns the state hash of cryptocurrency service.
     pub fn state_hash(&self) -> Vec<Hash> {
-        vec![self.wallets().object_hash()]
+        vec![self.wallets.object_hash()]
+    }
+}
+
+impl<T> Schema<T>
+where
+    T: AccessExt,
+    T::Base: IndexAccessMut,
+{
+    pub(crate) fn initialize(access: T) -> Self {
+        Self {
+            wallets: access.ensure_proof_map("wallets"),
+            access,
+        }
+    }
+
+    fn ensure_wallet_history(&self, public_key: &PublicKey) -> ProofListIndex<T::Base, Hash> {
```

Comment:
> it can be a structure field (when should it? when should it not?)

Top-level / singular indexes can be fields in the schema; grouped indexes cannot, at least with the current set of abstractions.

> what does `None` mean? a DB failure, forgotten initialization, empty index, something else?

`None` means no index in the database.

> why can't we always call `ensure_*` method?

Because `ensure_*` is only defined for writeable `AccessExt` implementations (e.g., `&Fork` or `Prefixed<&Fork>`). Readonly implementations naturally cannot perform write operations on the storage, such as creating indexes. Hence, I'd suggest to view `ensure_*` methods of `AccessExt` as ones that *ensure* that an index is there (these methods are idempotent, they do not create indexes on each call). Other methods *get* an index handle from the storage and return `None` if the index is not in the storage. Essentially, `access.ensure_something(addr)` is a shortcut for `access.something(addr).unwrap_or_else(|| Something::create(addr))`, except for now `Something::create` constructors are not publicly exposed.